File: linuxpict/render.py
from __future__ import annotations

import math

import cairo

from .model import Annotation, Rect, Tool


def draw_annotation(ctx: cairo.Context, item: Annotation) -> None:
    x1, y1 = item.start
    x2, y2 = item.end
    ctx.save()
    ctx.set_source_rgba(*item.color)
    ctx.set_line_width(item.width)
    ctx.set_line_cap(cairo.LINE_CAP_ROUND)
    ctx.set_line_join(cairo.LINE_JOIN_ROUND)
    if item.tool == Tool.LINE:
        ctx.move_to(x1, y1)
        ctx.line_to(x2, y2)
        ctx.stroke()
    elif item.tool == Tool.ARROW:
        ctx.move_to(x1, y1)
        ctx.line_to(x2, y2)
        angle = math.atan2(y2 - y1, x2 - x1)
        head = max(14, item.width * 3)
        for offset in (-0.55, 0.55):
            ctx.move_to(x2, y2)
            ctx.line_to(x2 - head * math.cos(angle + offset), y2 - head * math.sin(angle + offset))
        ctx.stroke()
    elif item.tool == Tool.BOX:
        rect = Rect.between(x1, y1, x2, y2)
        ctx.rectangle(rect.x, rect.y, rect.width, rect.height)
        ctx.stroke()
    elif item.tool == Tool.ELLIPSE:
        rect = Rect.between(x1, y1, x2, y2)
        ctx.translate(rect.x + rect.width / 2, rect.y + rect.height / 2)
        ctx.scale(max(rect.width / 2, 0.001), max(rect.height / 2, 0.001))
        ctx.arc(0, 0, 1, 0, math.tau)
        ctx.set_line_width(item.width / max(rect.width / 2, rect.height / 2, 0.001))
        ctx.stroke()
    elif item.tool == Tool.TEXT and item.text:
        ctx.select_font_face("Sans", cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_BOLD)
        ctx.set_font_size(max(16, item.width * 4))
        for line_no, line in enumerate(item.text.splitlines()):
            ctx.move_to(x1, y1 + max(16, item.width * 4) * (line_no + 1))
            ctx.show_text(line)
    ctx.restore()


def render_png(source_path: str, output_path: str, annotations: tuple[Annotation, ...], crop: Rect) -> None:
    source = cairo.ImageSurface.create_from_png(source_path)
    width, height = int(crop.width), int(crop.height)
    target = cairo.ImageSurface(cairo.FORMAT_ARGB32, width, height)
    ctx = cairo.Context(target)
    ctx.set_source_surface(source, -crop.x, -crop.y)
    ctx.paint()
    ctx.translate(-crop.x, -crop.y)
    for item in annotations:
        draw_annotation(ctx, item)
    target.write_to_png(output_path)

File: src/render.cpp
#include "linuxpict/render.hpp"

#include <cairomm/surface.h>

#include <algorithm>
#include <cmath>
#include <numbers>
#include <stdexcept>

namespace linuxpict {

void draw_annotation(const Cairo::RefPtr<Cairo::Context>& context, const Annotation& item) {
    context->save();
    context->set_source_rgba(item.red, item.green, item.blue, item.alpha);
    context->set_line_width(item.width);
    context->set_line_cap(Cairo::LINE_CAP_ROUND);
    context->set_line_join(Cairo::LINE_JOIN_ROUND);

    const auto x1 = item.start.x;
    const auto y1 = item.start.y;
    const auto x2 = item.end.x;
    const auto y2 = item.end.y;
    switch (item.tool) {
        case Tool::Line:
            context->move_to(x1, y1);
            context->line_to(x2, y2);
            context->stroke();
            break;
        case Tool::Arrow: {
            context->move_to(x1, y1);
            context->line_to(x2, y2);
            const auto angle = std::atan2(y2 - y1, x2 - x1);
            const auto head = std::max(14.0, item.width * 3);
            for (const auto offset : {-0.55, 0.55}) {
                context->move_to(x2, y2);
                context->line_to(
                    x2 - head * std::cos(angle + offset),
                    y2 - head * std::sin(angle + offset)
                );
            }
            context->stroke();
            break;
        }
        case Tool::Box: {
            const auto rect = Rect::between(item.start, item.end);
            context->rectangle(rect.x, rect.y, rect.width, rect.height);
            context->stroke();
            break;
        }
        case Tool::Ellipse: {
            const auto rect = Rect::between(item.start, item.end);
            context->translate(rect.x + rect.width / 2, rect.y + rect.height / 2);
            const auto radius_x = std::max(rect.width / 2, 0.001);
            const auto radius_y = std::max(rect.height / 2, 0.001);
            context->scale(radius_x, radius_y);
            context->arc(0, 0, 1, 0, std::numbers::pi * 2);
            context->set_line_width(item.width / std::max(radius_x, radius_y));
            context->stroke();
            break;
        }
        case Tool::Text:
            if (!item.text.empty()) {
                context->select_font_face("Sans", Cairo::FONT_SLANT_NORMAL, Cairo::FONT_WEIGHT_BOLD);
                const auto size = item.width <= 4 ? 24.0 : item.width <= 8 ? 36.0 : 56.0;
                context->set_font_size(size);
                std::size_t start = 0;
                int line = 1;
                while (start <= item.text.size()) {
                    const auto end = item.text.find('\n', start);
                    context->move_to(x1, y1 + size * line++);
                    context->show_text(item.text.substr(start, end - start));
                    if (end == std::string::npos) {
                        break;
                    }
                    start = end + 1;
                }
            }
            break;
        case Tool::Crop:
            break;
    }
    context->restore();
}

void render_png(
    const std::filesystem::path& source,
    const std::filesystem::path& destination,
    const std::vector<Annotation>& annotations,
    const Rect& crop
) {
    auto input = Cairo::ImageSurface::create_from_png(source.string());
    if (!input) {
        throw std::runtime_error("Could not decode captured PNG");
    }
    auto output = Cairo::ImageSurface::create(
        Cairo::FORMAT_ARGB32,
        static_cast<int>(crop.width),
        static_cast<int>(crop.height)
    );
    auto context = Cairo::Context::create(output);
    context->set_source(input, -crop.x, -crop.y);
    context->paint();
    context->translate(-crop.x, -crop.y);
    for (const auto& annotation : annotations) {
        draw_annotation(context, annotation);
    }
    output->write_to_png(destination.string());
}

}  // namespace linuxpict
